packages/terra-image/src/Image.jsx

```diff
@@ -47,36 +51,113 @@ const defaultProps = {
-  onLoad,
-  onError,
-  ...customProps
-}) => {
-  const classes = cx([
-    'image',
-    variant,
-    { fluid: isFluid },
-    customProps.className,
-  ]);
-
-  return (
-    <img
-      src={src}
-      alt={alt}
-      height={height}
-      width={width}
-      onLoad={onLoad}
-      onError={onError}
-      {...customProps}
-      className={classes}
-    />);
-};
+class Image extends React.Component {
+
+  constructor(props) {
+    super(props);
+
+    this.state = { isLoading: true, isError: false };
+
+    this.handleOnLoad = this.handleOnLoad.bind(this);
+    this.handleOnError = this.handleOnError.bind(this);
+  }
+
+  componentWillReceiveProps(newProps) {
+    // If a new image is being loaded, reset the state to loading
+    if (newProps.src !== this.props.src || newProps.placeholderSrc !== this.props.placeholderSrc) {
+      this.setState({ isLoading: true, isError: false });
+    }
+  }
+
+  handleOnLoad() {
+    this.setState({ isLoading: false });
+    const onLoad = this.props.onLoad;
+
+    if (onLoad !== undefined) {
+      onLoad();
+    }
+  }
+
+  handleOnError() {
+    this.setState({ isLoading: false, isError: true });
+    const onError = this.props.onError;
+
+    if (onError !== undefined) {
+      onError();
+    }
+  }
+
+  createPlaceholderImage(customProps, imageClasses) {
+    const { placeholderSrc, alt, height, width } = this.props;
+    return (
+      <img
+        src={placeholderSrc}
+        alt={alt}
+        height={height}
+        width={width}
+        className={imageClasses}
+        {...customProps}
+      />
+    );
+  }
+
+  createImage(customProps, imageClasses) {
+    const { src, alt, height, width } = this.props;
+    return (
+      <img
+        src={src}
+        alt={alt}
+        height={height}
+        width={width}
+        onLoad={this.handleOnLoad}
+        onError={this.handleOnError}
+        className={imageClasses}
+        {...customProps}
+      />
+    );
+  }
+
+  render() {
+    const { src, variant, isFluid, alt, placeholderSrc, height, width, onLoad, onError, ...customProps } = this.props;
+
+    const imageClasses = cx([
+      'image',
+      variant,
+      customProps.className,
+      { fluid: isFluid },
+    ]);
+    delete customProps.className;
+
+    if (placeholderSrc) {
+      if (this.state.isLoading) {
+        return (
+          <div>
+            <div className={cx('hidden')}>{this.createImage(customProps, imageClasses)}</div>
+            <div>{this.createPlaceholderImage(customProps, imageClasses)}</div>
+          </div>
+        );
+      }
+
+      return (
+        <div>
+          <div>
+            {
+              this.state.isError ?
+                this.createPlaceholderImage(customProps, imageClasses) :
+                this.createImage(customProps, imageClasses)
+            }
+          </div>
+        </div>
+      );
+    }
+
+    return (
+      <div>
+        <div>{this.createImage(customProps, imageClasses)}</div>
+      </div>
```

Comment:
Why are both the error and this render double nested with divs?

Reply:
That was how the old ProfileImage was written, where this code is reimagined from. I have removed them